cns/service_test.go

```diff
@@ -0,0 +1,304 @@
+		svc.SetOption(acn.OptCnsPort, "")
+
+		err = svc.Initialize(config)
+		t.Cleanup(func() {
+			svc.Uninitialize()
+		})
+		require.NoError(t, err)
+
+		err = svc.StartListener(config)
+		require.NoError(t, err)
+
+		mTLSConfig, err := getTLSConfigFromFile(config.TLSSettings)
+		require.NoError(t, err)
+
+		client := &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: mTLSConfig,
+			},
+		}
+
+		// TLS listener
+		req, err := http.NewRequestWithContext(context.TODO(), http.MethodGet, "https://localhost:10091", http.NoBody)
+		require.NoError(t, err)
+		resp, err := client.Do(req)
+		t.Cleanup(func() {
+			resp.Body.Close()
+		})
+		require.NoError(t, err)
+
+		// HTTP listener
+		httpClient := &http.Client{}
+		req, err = http.NewRequestWithContext(context.TODO(), http.MethodGet, "http://localhost:10090", http.NoBody)
+		require.NoError(t, err)
+		resp, err = httpClient.Do(req)
+		t.Cleanup(func() {
+			resp.Body.Close()
+		})
+		require.NoError(t, err)
+	})
+}
+
+func TestMtlsRootCAsFromCertificate(t *testing.T) {
+	testCertFilePath := createTestCertificate(t)
+
+	tlsSettings := serverTLS.TlsSettings{
+		TLSCertificatePath: testCertFilePath,
+	}
+	tlsCertRetriever, err := serverTLS.GetTlsCertificateRetriever(tlsSettings)
+	require.NoError(t, err)
+
+	cert, err := tlsCertRetriever.GetCertificate()
+	require.NoError(t, err)
+
+	key, err := tlsCertRetriever.GetPrivateKey()
+	require.NoError(t, err)
+
+	t.Run("returns root CA pool when provided a single self-signed CA cert", func(t *testing.T) {
+		// one root CA
+		tlsCert := tls.Certificate{
+			Certificate: [][]byte{cert.Raw},
+			PrivateKey:  key,
+			Leaf:        cert,
+		}
+
+		var r *x509.CertPool
+		r, err = mtlsRootCAsFromCertificate(&tlsCert)
+		require.NoError(t, err)
+		assert.NotNil(t, r)
+	})
+	t.Run("returns root CA pool when provided with a full cert chain", func(t *testing.T) {
+		// simulate a full cert chain (leaf cert + root CA cert)
+		tlsCert := tls.Certificate{
+			Certificate: [][]byte{cert.Raw, cert.Raw},
+			PrivateKey:  key,
+			Leaf:        cert,
+		}
+		require.NoError(t, err)
+		r, err := mtlsRootCAsFromCertificate(&tlsCert)
+		require.NoError(t, err)
+		assert.NotNil(t, r)
+	})
+	t.Run("does not return root CA pool when provided with no cert", func(t *testing.T) {
+		r, err := mtlsRootCAsFromCertificate(nil)
+		require.Error(t, err)
+		assert.Nil(t, r)
+
+		r, err = mtlsRootCAsFromCertificate(&tls.Certificate{})
+		require.Error(t, err)
+		assert.Nil(t, r)
+	})
+	t.Run("does not return root CA pool when provided with invalid certs", func(t *testing.T) {
+		tt := []struct {
+			invalidCert [][]byte
+		}{
+			{[][]byte{[]byte("invalid leaf cert")}},
+			{[][]byte{[]byte("invalid leaf cert"), []byte("invalid root CA cert")}},
+		}
+
+		for _, tc := range tt {
+			cert := tls.Certificate{
+				Certificate: tc.invalidCert,
+			}
+			r, err := mtlsRootCAsFromCertificate(&cert)
+			require.Error(t, err)
+			assert.Nil(t, r)
+		}
+	})
```

Comment:
These tests look good, but I think that it would be easier to read and add additional test cases later if they were table-driven instead of separate `test.Run`s

Reply:
refactored

Reply:
I really meant all of the new tests. But like I said, not a blocker

Reply:
ah gotcha, i will keep the `Service` tests as is. if there are newer cases i'll consider refactoring them